File: app/src/main/java/com/example/recyclerview/widget/UnderlineButton.java
package com.example.recyclerview.widget;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Typeface;
import android.util.AttributeSet;
import androidx.appcompat.widget.AppCompatTextView;

public class UnderlineButton extends AppCompatTextView {

    private final Paint underlinePaint;
    private boolean isSelectedState = false;

    private int dpToPx(float dp) {
        return (int) (dp * getResources().getDisplayMetrics().density + 0.5f);
    }

    public UnderlineButton(Context context) {
        this(context, null);
    }

    public UnderlineButton(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public UnderlineButton(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);

        setBackgroundColor(Color.WHITE);
        setPadding(dpToPx(16), dpToPx(12), dpToPx(16), dpToPx(12)); 
        setGravity(android.view.Gravity.CENTER);
        setBackground(null);
        setClickable(true);
        setTypeface(null, Typeface.NORMAL);

        underlinePaint = new Paint(Paint.ANTI_ALIAS_FLAG);
        underlinePaint.setStyle(Paint.Style.STROKE);
        underlinePaint.setStrokeWidth(dpToPx(2)); 
        underlinePaint.setColor(Color.parseColor("#A0E6B7"));
        underlinePaint.setStrokeCap(Paint.Cap.ROUND);
    }

    public void setCustomSelected(boolean selected) {
        if (this.isSelectedState != selected) {
            this.isSelectedState = selected;
            setTypeface(null, selected ? Typeface.BOLD : Typeface.NORMAL);
            invalidate();
        }
    }

    public boolean isCustomSelected() {
        return isSelectedState;
    }


    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);

        if (!isSelectedState) return;

        String text = getText().toString().trim();
        if (text.isEmpty()) return;

        Paint paint = getPaint();
        float textWidth = paint.measureText(text);
        Paint.FontMetrics fm = paint.getFontMetrics();

        float underlineY = getBaseline() + fm.descent + 0.7f * (fm.descent - fm.ascent);


        float avgCharWidth = textWidth / Math.max(1, text.length());
        float underlineWidth = Math.max(textWidth - avgCharWidth, dpToPx(8)); 


        int paddingLeft = getCompoundPaddingLeft();
        int paddingRight = getCompoundPaddingRight();
        float availableWidth = getWidth() - paddingLeft - paddingRight;
        float textCenterX = paddingLeft + availableWidth / 2f; 


        float startX = textCenterX - underlineWidth / 2f;
        float endX   = textCenterX + underlineWidth / 2f;

        canvas.drawLine(startX, underlineY, endX, underlineY, underlinePaint);
    }
}

File: app/src/main/java/com/example/recyclerview/widget/UnderlineButton.kt
package com.example.recyclerview.widget

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Typeface
import android.util.AttributeSet
import android.view.Gravity
import androidx.appcompat.widget.AppCompatTextView
import kotlin.math.max

class UnderlineButton @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : AppCompatTextView(context, attrs, defStyleAttr) {
    private val underlinePaint: Paint
    private var isSelectedState = false

    private fun dpToPx(dp: Float): Int {
        return (dp * getResources().getDisplayMetrics().density + 0.5f).toInt()
    }

    init {
        setBackgroundColor(Color.WHITE)
        setPadding(dpToPx(16f), dpToPx(12f), dpToPx(16f), dpToPx(12f))
        setGravity(Gravity.CENTER)
        setBackground(null)
        setClickable(true)
        setTypeface(null, Typeface.NORMAL)

        underlinePaint = Paint(Paint.ANTI_ALIAS_FLAG)
        underlinePaint.setStyle(Paint.Style.STROKE)
        underlinePaint.setStrokeWidth(dpToPx(2f).toFloat())
        underlinePaint.setColor(Color.parseColor("#A0E6B7"))
        underlinePaint.setStrokeCap(Paint.Cap.ROUND)
    }

    var isCustomSelected: Boolean
        get() = isSelectedState
        set(selected) {
            if (this.isSelectedState != selected) {
                this.isSelectedState = selected
                setTypeface(null, if (selected) Typeface.BOLD else Typeface.NORMAL)
                invalidate()
            }
        }


    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)

        if (!isSelectedState) return

        val text = getText().toString().trim { it <= ' ' }
        if (text.isEmpty()) return

        val paint: Paint = getPaint()
        val textWidth = paint.measureText(text)
        val fm = paint.getFontMetrics()

        val underlineY = getBaseline() + fm.descent + 0.7f * (fm.descent - fm.ascent)


        val avgCharWidth = textWidth / max(1, text.length)
        val underlineWidth = max(textWidth - avgCharWidth, dpToPx(8f).toFloat())


        val paddingLeft = getCompoundPaddingLeft()
        val paddingRight = getCompoundPaddingRight()
        val availableWidth = (getWidth() - paddingLeft - paddingRight).toFloat()
        val textCenterX = paddingLeft + availableWidth / 2f


        val startX = textCenterX - underlineWidth / 2f
        val endX = textCenterX + underlineWidth / 2f

        canvas.drawLine(startX, underlineY, endX, underlineY, underlinePaint)
    }
}